### PR TITLE
Phase 5: Curriculum Learning + Wider Transolver — Sample Difficulty Scheduling (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Combine **curriculum learning** (train on easy samples first, progressively introduce harder ones) with a **wider Transolver** (n_hidden=256). Phase 5 round 1 showed that all new architectures failed — the Transolver is fundamentally the right architecture. The question is how to extract more from it.

**Why this should work:**

**Curriculum Learning:**
- With 1,322 training samples spanning easy (single-foil, in-distribution) to hard (tandem, extreme conditions), random sampling mixes difficulty levels. Starting with easier samples and gradually introducing harder ones helps the model build robust representations before tackling the challenging cases.
- Easy first: single-foil racecar (simple geometry, moderate Re), then tandem racecar (interference), then cruise (high Re, extreme AoA).
- The WeightedRandomSampler already balances domains — curriculum extends this to difficulty within each domain.

**Wider Model:**
- Current model uses 36/96 GB VRAM. A wider model (n_hidden=256) uses more capacity without changing the architecture.
- The tanjiro Scale Up experiment (#1884) is testing this independently. Here we combine width with curriculum.

### Curriculum Implementation:
Add `--curriculum` flag. When set, implement sample difficulty scoring and scheduled weighting:

```python
def compute_difficulty(x, y, model, stats):
    """Score sample difficulty by model's current loss on that sample."""
    with torch.no_grad():
        pred = model(x)
        loss = F.l1_loss(pred, y, reduction='none').mean(dim=(1,2))
    return loss  # higher = harder

# In training loop, adjust sampler weights based on epoch:
if cfg.curriculum:
    curriculum_progress = min(1.0, epoch / cfg.curriculum_epochs)
    # Start with uniform, progressively upweight harder samples
    difficulty_weight = 1.0 + curriculum_progress * (sample_difficulty - sample_difficulty.mean())
    sampler.weights = base_weights * difficulty_weight.clamp(min=0.1)
```

Alternatively (simpler): sort training samples by domain difficulty and introduce domains progressively:
- Epochs 0-30: only racecar_single
- Epochs 30-60: + racecar_tandem
- Epochs 60+: + cruise (full training set)

### Hyperparameters across 8 GPUs:

| GPU | n_hidden | curriculum | curriculum_type | Notes |
|-----|----------|-----------|----------------|-------|
| 0 | 256 | domain_curriculum | progressive domains | Wider + domain curriculum |
| 1 | 192 | domain_curriculum | progressive domains | Standard width + domain curriculum |
| 2 | 256 | loss_curriculum | reweight by difficulty | Wider + loss-based curriculum |
| 3 | 256 | none | — | Just wider (no curriculum) |
| 4 | 192 | loss_curriculum | reweight by difficulty | Standard + loss-based curriculum |
| 5 | 256 | anti_curriculum | hard first | Wider + anti-curriculum (control) |
| 6 | 256 | domain_curriculum | progressive domains | seed 43 |
| 7 | 192 | none | — | **Baseline** |

```bash
BASE="--agent askeladd --field_decoder --adaln_output --use_lion --lr 2e-4 --cosine_T_max 180 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --disable_pcgrad --pressure_first --pressure_deep --seed 42"

CUDA_VISIBLE_DEVICES=0 python train.py $BASE --wandb_name "askeladd/p5-w256-domcurr" --wandb_group phase5-curriculum --n_hidden 256 --curriculum domain
# GPU 7: Baseline
CUDA_VISIBLE_DEVICES=7 python train.py $BASE --wandb_name "askeladd/p5-baseline" --wandb_group phase5-curriculum
```

## Baseline
val/loss 0.401, p_in 12.95, p_oodc 8.40, p_tan 33.8, p_re 24.7 (PR #1867)

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: ## Phase 5 Results — Curriculum Learning + Wider Model

### Final Results

| Run | val/loss | p_in | p_oodc | p_re | p_tan |
|-----|----------|------|--------|------|-------|
| **Baseline** | **0.399** | **12.6** | 8.4 | 24.6 | **32.9** |
| w192-losscurr | 0.405 | 13.7 | 8.2 | 24.8 | 33.6 |
| w192-domcurr | 0.407 | — | — | — | — |
| w256-domcurr | 0.436 | — | — | — | — |

### Decision: **CLOSE** — Curriculum learning provides no improvement

Domain curriculum and loss-based curriculum both converge to the same result as baseline. The WeightedRandomSampler already balances domains effectively, and the model doesn't benefit from progressive difficulty scheduling. Wider models (w256) are consistently worse due to overfitting on 1,322 samples.

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

All 8 experiments completed (~180 min each, wall-clock limited).

### Summary Table

| GPU | Config | Epochs | val/loss | p_in | p_tan | p_oodc | p_re | Peak VRAM | W&B Run |
|-----|--------|--------|----------|------|-------|--------|------|-----------|---------|
| 0 | 256 + domain curriculum | 137 | 0.443 | 15.6 | 34.4 | 11.1 | 26.0 | 46.1GB | gtt5e8ii |
| 1 | 192 + domain curriculum | 171 | **0.406** | **13.8** | **33.2** | 8.8 | 24.9 | 36.2GB | xmhm6cc2 |
| 2 | 256 + loss curriculum | 121 | 0.457 | 18.9 | 35.7 | 11.1 | 25.9 | 46.4GB | pbb5inpb |
| 3 | 256 only (no curriculum) | 123 | 0.457 | 17.0 | 36.4 | 10.2 | 26.4 | 46.1GB | 1b2jnp6b |
| 4 | 192 + loss curriculum | 156 | **0.405** | 13.7 | 33.6 | **8.2** | **24.8** | 36.5GB | b519ipvy |
| 5 | 256 + anti-curriculum | 120 | 0.445 | 16.1 | 35.3 | 10.0 | 25.8 | 46.4GB | ket8zqth |
| 6 | 256 + domain curriculum (seed 43) | 137 | 0.433 | 14.5 | 34.2 | 10.8 | 26.4 | 46.1GB | oi8ieiu0 |
| 7 | **192 baseline** (no curriculum) | 159 | **0.399** | **12.6** | **32.9** | **8.4** | **24.6** | 36.2GB | tnyqne92 |

### Baseline comparison (PR #1867)

| Metric | Baseline (this run) | Baseline (PR body) | Best curriculum (GPU 4) | Delta |
|--------|--------------------|--------------------|------------------------|-------|
| val/loss | 0.399 | 0.401 | 0.405 | +1.5% |
| p_in | 12.6 | 12.95 | 13.7 | +9% |
| p_oodc | 8.4 | 8.40 | **8.2** | **-2.4%** |
| p_tan | 32.9 | 33.8 | 33.6 | +2.1% |
| p_re | 24.6 | 24.7 | 24.8 | +0.8% |

### Commands used

```bash
BASE="--agent askeladd --field_decoder --adaln_output --use_lion --lr 2e-4 --cosine_T_max 180 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --disable_pcgrad --pressure_first --pressure_deep --seed 42"

# GPU 0: Wider + domain curriculum
CUDA_VISIBLE_DEVICES=0 python train.py $BASE --wandb_name "askeladd/p5-w256-domcurr" --wandb_group phase5-curriculum --n_hidden 256 --curriculum domain
# GPU 1: Standard + domain curriculum
CUDA_VISIBLE_DEVICES=1 python train.py $BASE --wandb_name "askeladd/p5-w192-domcurr" --wandb_group phase5-curriculum --curriculum domain
# GPU 2: Wider + loss curriculum
CUDA_VISIBLE_DEVICES=2 python train.py $BASE --wandb_name "askeladd/p5-w256-losscurr" --wandb_group phase5-curriculum --n_hidden 256 --curriculum loss
# GPU 3: Wider only
CUDA_VISIBLE_DEVICES=3 python train.py $BASE --wandb_name "askeladd/p5-w256-nocurr" --wandb_group phase5-curriculum --n_hidden 256
# GPU 4: Standard + loss curriculum
CUDA_VISIBLE_DEVICES=4 python train.py $BASE --wandb_name "askeladd/p5-w192-losscurr" --wandb_group phase5-curriculum --curriculum loss
# GPU 5: Wider + anti-curriculum
CUDA_VISIBLE_DEVICES=5 python train.py $BASE --wandb_name "askeladd/p5-w256-anticurr" --wandb_group phase5-curriculum --n_hidden 256 --curriculum anti
# GPU 6: Wider + domain curriculum (seed 43)
CUDA_VISIBLE_DEVICES=6 python train.py $BASE --wandb_name "askeladd/p5-w256-domcurr-s43" --wandb_group phase5-curriculum --n_hidden 256 --curriculum domain --seed 43
# GPU 7: Baseline
CUDA_VISIBLE_DEVICES=7 python train.py $BASE --wandb_name "askeladd/p5-baseline-curriculum" --wandb_group phase5-curriculum
```

### What happened

**Neither curriculum learning nor wider models beat the baseline**, though loss-based curriculum (GPU 4) came closest and achieved the **best OOD-condition pressure MAE (8.2 vs baseline 8.4)**.

**Key findings:**

1. **Curriculum adds no clear benefit at standard width**: GPU 1 (192 + domain curriculum, val/loss=0.406) and GPU 4 (192 + loss curriculum, val/loss=0.405) both marginally trail the baseline (0.399). The gap is within noise.

2. **Wider model (256) hurts more than helps**: All n_hidden=256 variants (GPUs 0, 2, 3, 5, 6) performed significantly worse than the baseline. The wider model gets ~30% fewer epochs (120-137 vs 159-171) due to ~87s/epoch vs ~67s for 192. The extra capacity doesn't compensate for fewer optimization steps.

3. **Domain curriculum disrupts OOD learning**: GPUs 0 and 6 (wider + domain) ended with worse OOD-cond (11.1, 10.8) than the wider-only control GPU 3 (10.2). Withholding cruise data until epoch 40 means the model has far fewer gradient updates on cruise samples by convergence.

4. **Loss-based curriculum helps OOD slightly**: GPU 4 achieved **p_oodc=8.2**, marginally beating both the baseline (8.4) and the loss curriculum PR benchmark (8.40). The progressive upweighting of hard samples appears to help with extreme conditions, though the overall effect is small.

5. **Anti-curriculum (GPU 5) was competitive**: p_oodc=10.0 at 256 width — better than random curriculum baselines at 256 width. Hard-first training may help wider models that are more prone to underfitting on hard samples early.

6. **The throughput bottleneck dominates**: The most important factor was total training iterations. Standard-width models (192) got 155-171 epochs, wider (256) got 120-137. With EMA starting at epoch 140, the wider models barely benefit from model averaging. This explains why all wider variants underperform.

### Suggested follow-ups

1. **Loss curriculum at standard width with longer training**: GPU 4 showed the best OOD-cond. Run it for 360 min to confirm whether the benefit holds or was noise.
2. **Curriculum with earlier EMA**: Since domain curriculum delays OOD data, the model needs more time to converge on those domains. Try `--ema_start_epoch 100` with domain curriculum.
3. **Wider model with reduced batch size**: The throughput penalty of n_hidden=256 is severe. Try batch_size=3 to maintain similar iteration count, or explore gradient checkpointing.
4. **Domain curriculum with shorter phases**: curriculum_epochs=30 instead of 60, so cruise data is introduced by epoch 20 instead of 40.
